Switching to using PHP-FPM

oparoz opened the pull request:
@kyrofa @jospoortvliet @ezraholm50 @enoch85 @miska

Snaps of this PR for your beta PiClouds are available from here:
https://code.launchpad.net/~olivier-interfasys/+snap/owncloud-snap

Sideloading instructions:
https://github.com/owncloud/pi-image/wiki/Sideloading-snaps-for-testing

I've added comments in-line to explain the various components.
### TODO
- [x] Remove `export OWNCLOUD_DATABASE_PASSWORD=$(cat $owncloud_password_path)`
- [x] Add PHP extensions to PHP cli. Can be another PR

oparoz commented on parts/plugins/x-apache.py:
```diff
@@ -185,7 +185,7 @@ def build(self):
         shutil.copytree(apache_source_directory, apache_build_directory)
 
-        subprocess.check_call("./configure --prefix={} --enable-modules=none --enable-mods-shared='{}' ENABLED_DSO_MODULES='{}'".format(self.installdir, ' '.join(self.options.modules), ','.join(self.options.modules)),
+        subprocess.check_call("./configure --prefix={} --with-mpm=event --enable-modules=none --enable-mods-shared='{}' ENABLED_DSO_MODULES='{}'".format(self.installdir, ' '.join(self.options.modules), ','.join(self.options.modules)),
```
Force event MPM to lower memory usage.

oparoz commented on parts/plugins/x-apache.py:
```diff
@@ -247,7 +247,6 @@ def _build_third_party_modules(self):
 
             configure_command = [
                 './configure', '--prefix=' + self.installdir,
-                '--with-apxs2={}/bin/apxs'.format(self.installdir),
```
Remove mod_php

kyrofa replied:
Yeah you should probably not cannibalize the apache plugin here-- FPM doesn't depend on apache, so it doesn't belong here.

oparoz replied:
Yeah, will give it a go

oparoz commented on snapcraft.yaml:
```diff
@@ -1,8 +1,7 @@
 version: 9.0.0ubuntu1
 summary: ownCloud
-description: ownCloud running on Apache with MySQL. This is currently in beta.
-icon: icon.svg
```
The icon currently gets in the way, will probably be restored before merging

kyrofa replied:
The bug that you were hitting has been fixed in the most recent release of Snapcraft (2.7). Please update when you're able and put this back.

oparoz replied:
Thanks

oparoz commented on snapcraft.yaml:
```diff
@@ -12,6 +11,13 @@ apps:
     daemon: simple
     plugs: [listener]
 
+  # PHP-FPM daemon
```
That's the daemon responsible for keeping PHP-FPM alive and ready to process requests

oparoz commented on snapcraft.yaml:
```diff
@@ -45,7 +51,8 @@ parts:
       - headers
-      - fcgid
+      - proxy
+      - proxy_fcgi
```
We use mod_proxy_fcgi now

oparoz commented on snapcraft.yaml:
```diff
@@ -63,20 +70,29 @@ parts:
+          - --enable-fpm
+          - --disable-pear
+          - --disable-phar
+          - --disable-phpdbg
```
Enabling PHP-FPM, disabling unneeded things.
People won't be installing libs via PEAR or launch PHAR executables
We use xdebug and not phpdbg

oparoz commented on snapcraft.yaml:
```diff
@@ -63,20 +70,29 @@ parts:
           - --with-gd
+          - --enable-gd-native-ttf
+          - --with-jpeg-dir=/usr/lib
+          - --with-freetype-dir=/usr/lib
```
That's issue #26 already fixed. Can be easily removed

kyrofa replied:
Though that fix still needs to be proposed.

oparoz commented on snapcraft.yaml:
```diff
@@ -63,20 +70,29 @@ parts:
           - --with-mcrypt
           - --enable-exif
+          - --enable-pcntl
+          - --enable-intl
```
Some missing extensions, more to come in other PRs

oparoz commented on snapcraft.yaml:
```diff
@@ -88,18 +104,25 @@ parts:
+      src/php/php.ini: php/lib/php.ini
+      src/php/extensions.ini: php/lib/php.conf.d/extensions.ini
+      src/php/php-fpm.conf: php/etc/php-fpm.conf
+      src/php/php-fpm: bin/php-fpm
```
New config files and the init script for php-fpm

kyrofa replied:
As of Snapcraft v2.7 you can leave off the destination file names if you don't want to rename them, so this becomes:

``` yaml
src/php/php.ini: php/lib/
src/php/extensions.ini: php/lib/php.conf.d/
# ...
```

Less typing, less error-prone.

oparoz replied:
That's good news :)

oparoz commented on src/owncloud/apache_config:
```diff
@@ -1,9 +1,10 @@
+SetEnvIfNoCase ^Authorization$ "(.+)" HTTP_AUTHORIZATION=$1
+ProxyTimeout 900
+<FilesMatch "\.(inc|php)$">
+	SetHandler "proxy:unix:${SNAP_DATA}/php/sockets/www-data.sock|fcgi://localhost"
```
Using mod_proxy_fcgi to connect Apache and PHP-FPM

oparoz commented on src/owncloud/apache_config:
```diff
@@ -14,3 +15,14 @@ Alias "/apps" "${SNAP_DATA}/owncloud/apps"
+	ServerLimit				4
+	MaxRequestWorkers		64
+	MaxConnectionsPerChild	0
+</IfModule>
```
An example mpm_event config, can easily be modified here or in another PR

kyrofa replied:
Yeah let's remove this here. This rpi-specific stuff will probably need to go in the fork.

oparoz replied:
Actually, I had forgotten that this only works with PHP-FPM, so it's best to keep a sane starting point here, which can be tuned later, depending on when and if we agree on a better config.

kyrofa replied:
Just so I understand, are you saying that php-fpm _requires_ this?

oparoz replied:
It's not a hard requirement, but they go hand in hand. You don't get all the benefits of using PH-FPM without using mpm_event.

kyrofa replied:
Okay. Just try to keep configs pretty general in this repo-- we can really tune it for the rpi in the fork. I'll trust you here, since I know nothing about these values :wink: .

oparoz commented on src/owncloud/autoconfig.php:
```diff
@@ -3,23 +3,26 @@
-$AUTOCONFIG = array(
-'directory' => $data_path.'/owncloud/data',
+$writable = getenv('SNAP_DATA');
+$owncloud_password = trim(file_get_contents($writable . '/mysql/owncloud_password'));
```
Fetching the DB password from the file directly instead of using environment variables

kyrofa replied:
Why? Particularly since you didn't remove the variable itself?

oparoz replied:
Removing the variable is already on the todo, that env variable is not always available to PHP-FPM and can't be added afaik.

kyrofa replied:
Ah right, because now it's a separate process. Keep in mind as you refactor this that (especially on the pi) mysql takes quite some time to set itself up and generate the password upon first install. You don't want to evaluate this config file until that file exists.

oparoz replied:
From what I understand, Apache won't start before this is set up, so the user won't be able to open the page before that password is written to disk.

kyrofa replied:
Right, but if you remove the fact that the variable is defined there it's really waiting for a file to exist that it doesn't use. I wonder if there's a better place for it.

oparoz replied:
> it's really waiting for a file to exist that it doesn't use.

I don't understand. It does use the file

kyrofa replied:
> I don't understand. It does use the file

It does now, but only to define the variable. We'll just need to make sure we document what we're doing that so it doesn't appear crufty is all :smiley: . The only other place to put it is the php-fpm startup script, and I don't really feel like it belongs there either.

oparoz replied:
> It does now, but only to define the variable

I still don't get it :D. The variable is used when setting things up, then the file is destroyed and the variable doesn't exist any more anywhere.

> We'll just need to make sure we document what we're doing that so it doesn't appear crufty is all :smiley:

I'll add a comment.

> The only other place to put it is the php-fpm startup script, and I don't really feel like it belongs there either.

No it doesn't belong there. That file is loaded every time a PHP-FPM process is started.

kyrofa replied:
> I still don't get it :D. The variable is used when setting things up, then the file is destroyed and the variable doesn't exist any more anywhere.

I'm talking about [this](https://github.com/kyrofa/owncloud-snap/blob/master/src/owncloud/setup_owncloud#L12). All it uses the file for is to define that variable. If you remove the variable definition from it, then all you have left is "I'm waiting for this file.... okay move on." Though honestly that's no different than what we do for MySQL only a few lines up. The file isn't actually destroyed, and the variable is defined every time (so I quite like your change). I don't think the file should be destroyed, but the variable is definitely not necessary anymore.

oparoz replied:
Ah, it's clearer now, so I don't see a problem with having to wait for the password to be defined and saved to file in that `setup_owncloud` script. We have to wait for the DB to be set up anyway.

> The file isn't actually destroyed

I was talking about autoconfig.php

> but the variable is definitely not necessary anymore.

It's already remove in what I'll push later

oparoz commented on src/php/extensions.ini:
```diff
@@ -0,0 +1,5 @@
+; Extensions are located in $SNAP/lib/php/extensions/no-debug-non-zts-20151012
+; OpCache
+zend_extension=${SNAP}/lib/php/extensions/no-debug-non-zts-20151012/opcache.so
```
Loading opcache, may require some tuning, see #31

oparoz commented on src/php/php-fpm:
```diff
@@ -0,0 +1,172 @@
+	mkdir -p $PHPFOLDER/log
+	touch $PHPFOLDER/log/php-fpm.log
+	mkdir -p $PHPFOLDER/sockets
+fi
```
That setup routine could be placed in a separate script, if better in terms of management.

oparoz commented on src/php/php-fpm.conf:
```diff
@@ -0,0 +1,76 @@
+
+[www-data]
+user = root
+group = root
```
We have to use root!

oparoz commented on src/php/php-fpm:
```diff
@@ -0,0 +1,172 @@
+CONFINI=$CONFFOLDER/php/lib/php.ini
+PHP_INI_SCAN_DIR=$SNAP/php/lib/php.conf.d
+export PHP_INI_SCAN_DIR
+OPTS="-R --daemonize --pid=${PID} --fpm-config $FPMCONF -c $CONFINI"
```
`-R` allows us to run PHP-FPM as root.

oparoz commented on src/php/php-fpm.conf:
```diff
@@ -0,0 +1,76 @@
+pm = ondemand
+pm.max_children = 9
+pm.process_idle_timeout = 60
+pm.max_requests = 200
```
Can be tuned, see #28

oparoz commented on src/php/php.ini:
```diff
@@ -186,7 +186,7 @@
 ; Enable the PHP scripting language engine under Apache.
 ; http://php.net/engine
-engine = On
+engine = Off
```
Disable mod_php

oparoz commented on src/php/php.ini:
```diff
@@ -720,7 +720,7 @@ user_dir =
 ; Directory in which the loadable extensions (modules) reside.
 ; http://php.net/extension-dir
-; extension_dir = "./"
+extension_dir = "${SNAP}/lib/php/extensions/no-debug-non-zts-20151012"
```
Set the extension folder so that PHP-FPM can find our future extensions.

kyrofa replied:
`no-debug-non-zts-20151012` feels like a build artifact. Is it possible to use `${SNAP}/lib/php/extensions/` ?

oparoz replied:
It's actually a standard path, sort of like an ABI identifier.

kyrofa replied:
Ah, okay :smiley: .

oparoz commented:
Snaps of this PR for your beta PiClouds are available from here:
https://code.launchpad.net/~olivier-interfasys/+snap/owncloud-snap

oparoz commented:
It makes a huge difference on my device. 
If someone could look at theirs and give me the amount of RAM they have free when syncing.
This is what I have now:

```
$ vmstat 5
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 3  0      0 335956  75304 305636    0    0     0  1823 4691 7146 50 20 24  6  0
```

oparoz commented:
Sideloading instructions:
https://github.com/owncloud/pi-image/wiki/Sideloading-snaps-for-testing

jospoortvliet commented:
Oh halleluja, great work Oparoz. I will give this a test tomorrow, won't have time today. But rockingly cool!

kyrofa commented on src/php/php-fpm:
```diff
@@ -0,0 +1,172 @@
+export PHP_INI_SCAN_DIR
+OPTS="-R --daemonize --pid=${PID} --fpm-config $FPMCONF -c $CONFINI"
+
+#ulimit -HSn 32768
```
No need to include crufty comments.

oparoz replied:
How do we set that ulimit though?

kyrofa replied:
I doubt you can from within the snap. You typically can't affect anything system-wide.

kyrofa commented:
After you've finished refactoring this, please make the PR into develop instead of master (more information in the [contribution guide](https://github.com/kyrofa/owncloud-snap/blob/master/CONTRIBUTING.md)).

oparoz commented:
:+1:

Commit:
Increase max upload to 16GB. (#46)

This is done by using SNAP_DATA/owncloud/tmp instead of /tmp as
the upload_tmp_dir and also updating the upload_max_filesize and
upload_max_filesize in php.ini.

Fixes #5.

Signed-off-by: Kyle Fazzari <kyle@canonical.com>

oparoz committed:
Apache: Update to 2.4.20 (#50)

Fixes #48

oparoz committed:
Use PHP-FPM instead of mod_php
Fixes #39

oparoz commented:
Replaced by #52
